src/classes/CW_AudienceCSVUploadControllerTest.cls

```diff
@@ -0,0 +1,397 @@
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		String memebrsCSVString = 'Group Name,Account Name' + CSVReader.lineTerminator;
+		
+		for(Integer i=1; i<=1000; i++)
+		{
+			
+			String playerId = 'shl' + i;
+			memebrsCSVString += (i <= 500 ? group1 : group2) + ',' + playerId + CSVReader.lineTerminator;
+		}
+		memebrsCSVString = memebrsCSVString.removeEnd(CSVReader.lineTerminator);
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		uc.csvFile = Blob.valueOf(memebrsCSVString);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		uc.loadBatchJobProgressPage();
+		Test.stopTest();
+		system.assertEquals(1000,[select count() from Contact where Account_Name__c != :null]);
+		// 2 new groups were added
+		system.assertEquals(2,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+		system.assertEquals(500,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__r.Name = :group1]);
+		system.assertEquals(500,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__r.Name = :group2]);
+	}
+	
+	static testMethod void parseFaultyGroups()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		CW_Audience_Group__c audGr5 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr5.Name = group5;
+		audGr5.Is_Control_Group__c = true;
+		insert audGr5;
+		
+		CW_Audience_Group__c audGr1 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr1.Name = group1;
+		audGr1.Is_Control_Group__c = false;
+		audGr1.Control_Group__c = audGr5.id;
+		insert audGr1;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		Boolean exceptionHappend = false;
+		
+		String faultyCSVString = 'group name,account name' + CSVReader.lineTerminator;
+		
+		faultyCSVString += 'group3,' + CSVReader.lineTerminator;
+		faultyCSVString += group2 + ',333333';
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.Id;
+		//try to upload the file with blank fields
+		try{
+			uc.csvFile = Blob.valueOf(faultyCSVString);
+			uc.filename='FaultyFile.csv';
+			List<List<string>> csvLines = CSVReader.readCSVFile(faultyCSVString);
+		}
+		catch(Exception e)
+        {
+        	exceptionHappend = true;
+        }
+		Test.stopTest();
+		
+		// none of the groups or members gets uploaded if file contains null values
+		system.assertEquals(2,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.